packages/diagram-ui/src/features/toolbox/toolbox.ts

```diff
@@ -0,0 +1,548 @@
+        );
+    }
+
+    /**
+     * Generates the preview for a toolbox edit if available.
+     * If no preview is available, undefined is returned.
+     *
+     * @param toolboxEdit The toolbox edit
+     * @returns The preview or undefined
+     */
+    private generatePreviewIfAvailable(toolboxEdit: ToolboxEdit): VNode | undefined {
+        if (this.showPreviewFor != toolboxEdit.edit) {
+            return undefined;
+        }
+        const preview = this.elementPreviews.get(toolboxEdit.edit);
+        if (preview == undefined) {
+            return undefined;
+        }
+        return h(
+            "div.preview",
+            {
+                hook: {
+                    insert: (vnode) => {
+                        const element = vnode.elm as HTMLElement;
+                        const parent = element.parentElement!;
+                        const toolbox = parent.closest(".toolbox")!;
+                        const offset = parent.getBoundingClientRect().top - toolbox.getBoundingClientRect().top;
+                        element.style.top = `${offset}px`;
+                    }
+                }
+            },
+            preview
+        );
+    }
+
+    /**
+     * Extracts the toolbox edits from the root element.
+     *
+     * @param root The root element
+     * @returns The toolbox edits
+     */
+    private getToolboxEdits(root: Root): ToolboxEdit[] {
+        return Object.keys(root.edits)
+            .filter((key) => key.startsWith("toolbox/"))
+            .map((key) => {
+                const [group, name] = key.substring("toolbox/".length).split("/");
+                return { group, name, edit: key as `toolbox/${string}` };
+            });
+    }
+
+    /**
+     * Shows a preview for a toolbox edit.
+     * If no preview is available, a prediction is requested.
+     *
+     * @param edit The toolbox edit to show a preview for
+     */
+    private showPreview(edit: ToolboxEdit): void {
+        this.showPreviewFor = edit.edit;
+        if (!this.elementPreviews.has(edit.edit)) {
+            this.requestPrediction(edit);
+        } else {
+            this.update();
+        }
+    }
+
+    /**
+     * Requests a prediction for a toolbox edit.
+     *
+     * @param edit The toolbox edit to request a prediction for
+     */
+    private async requestPrediction(edit: ToolboxEdit): Promise<void> {
+        const request: ToolboxEditPredictionRequestAction = {
+            kind: ToolboxEditPredictionRequestAction.KIND,
+            edit: {
+                types: [edit.edit],
+                values: {
+                    x: 0,
+                    y: 0,
+                    prediction: true
+                },
+                elements: [this.currentRoot!.id]
+            },
+            requestId: generateRequestId()
+        };
+        const result = await this.actionDispatcher.request(request);
+        let preview: VNode | undefined = undefined;
+        if (result.root != undefined) {
```

Comment:
What we could do as well is allow passing an optional SVG as image to the edit.
If it is set, we use that instead.

Reply:
Yes, but this has several drawbacks
- it does not necessarily match what the change will look like, in particular if you use any styles
- light/dark mode will cause issues
- fonts will cause issues (these svg should contain no text)

So while possible, do we have a case where this improves the UX?

Reply:
We could display how messages will look like all of a sudden.

Reply:
Since at the moment, you need to have the Hylimo knowledge to know what it will look like.

Reply:
and this is not possible using predictions?

Reply:
Well… at least I don't see any association edits in the toolbox yet.